### ISSUE TITLE
Inconsistent behavior when returning from withContext and outer CoroutineContext has been cancelled

### ISSUE THREAD

**Issue by ashughes** [Kotlin/kotlinx.coroutines#1921]
The current implementation of `withContext` has inconsistent behavior when returning to an already cancelled outer `CoroutineContext`, depending on which path is taken (`FAST PATH #1`, `FAST PATH #2`, or `SLOW PATH`).

Both `FAST PATH #1` and `FAST PATH #2` will return the result of `withContext` to the outer `CoroutineContext` even if the outer `CoroutineContext` is already cancelled.

However, `SLOW PATH` will not return the result of `withContext` to the outer `CoroutineContext` if the outer `CoroutineContext` is cancelled. Instead, it throws a `CancellationException`. IMO, this is the expected behavior.

**Comment by elizarov** [comment]
Both fast path #1 and #2 return without actually performing any suspension in execution. The consistent approach we try to follow in `kotlinx.coroutines` (with as few exceptions as possible) is that all the suspending functions check for cancellation when execution actually suspends, (releasing and later reacquiring the current threads). Do be more specific, we make sure that cancelled coroutine does not grab a thread it was not previously occupying, but we make no effort to release the thread it holds while it was cancelled earlier.

What is your use-case for checking cancellation on fast return from `withContext`? What problem were you trying to solve when you stumbled upon this specific behavior?

**Comment by ashughes** [comment]
Thanks for the explanation. I just wanted to highlight the inconsistent behavior, which could lead to problems when switching dispatchers of either the outer `CoroutineContext` or the `withContext` parameter (which we may not have control over), e.g., a change in dispatcher could result in a change in behavior, which is not intuitive. It would be great if either `withContext` behaved consistently when returning to a cancelled coroutine or if the documentation was updated to describe the different behavior depending on the `CoroutineContext`s.

I discovered this while testing a bunch of different combinations of `withContext` usage. The motivation for this was trying to implement a good pattern for non-cancellable work. Originally I was using `withContext(NonCancellable)` but I came across [this article](https://medium.com/androiddevelopers/coroutines-patterns-for-work-that-shouldnt-be-cancelled-e26c40f142ad) and was exploring using an `applicationScope` instead. While testing this, I noticed that I get different behavior when the `applicationScope` is `GlobalScope` vs pretty much anything else. I wrote about that [here](https://medium.com/@ashughes/i-was-exploring-using-withcontext-as-described-in-what-about-something-simpler-3223bc1da4fe).

During this testing, I was also trying different dispatchers with different scopes. This is when I discovered the different behaviors for the paths. For my use case, it's not critical that `withContext` not return for fast path 1 and 2, but I thought it was inconsistent behavior. Plus, I originally misinterpreted the documentation:

> Note that the result of `withContext` invocation is dispatched into the original context in a cancellable way, which means that if the original [coroutineContext], in which `withContext` was invoked, is cancelled by the time its dispatcher starts to execute the code, it discards the result of `withContext` and throws [CancellationException].

I originally took "is cancelled by the time its dispatcher starts to execute the code" to mean the outer `CoroutineContext`. I now see from #1002 that this is referring to the completion check on the (new) `CoroutineContext` of the `withContext`.
